File: halatuju-web/src/lib/admin-api/requests.ts
/**
 * The Requests space (Sprint 15): a request, its conversation, its AI analysis, its quote and
 * schedule, and its screenshot attachments.
 *
 * ⚠ TWO SPANS of the old `admin-api.ts` (2140-2275 and 2698-2865). The whole billing and
 * invoicing module was written between the request TYPES and the calls that act on them; the
 * split puts the domain back together, which is one of the two the roadmap promised for free.
 */
import { adminFetch, adminMutate } from './client'
import type { ApiOptions } from './client'

// ---- Requests space (Sprint 15) ----
// Bug/feature requests → AI reviewer → owner-gated hours quotes. Dark behind REQUESTS_ENABLED:
// every route 404s while off, which is how the Administration hub card ships dark (the count
// probe 404s → the card is hidden; no client flag).

/**
 * One entry in a request's DISCUSSION (TD-201) — the replacement for the `clarifications`
 * question/answer pairs, which are no longer on either payload.
 *
 * A QUESTION IS A COMMENT AWAITING A REPLY (`awaiting_reply`), which is what makes this one
 * stream rather than a thread of questions with a separate comment log beside it.
 *
 * ⚠ `visibility: 'internal'` NEVER reaches an org_admin — the server filters the ROWS out of the
 * org payload (`org_requests.comments_for`). It appears in this type because the SUPER payload
 * carries both, and the detail page badges an internal comment so the owner can see at a glance
 * what the requester cannot. Do not treat its absence client-side as the protection; the
 * protection is server-side and tested there.
 */
export interface OrgRequestComment {
  id: number
  // WHO spoke. 'ai' is the reviewer and 'engineer' is the approved analysis (TD-204) — neither has
  // a PartnerAdmin row, so `author_name` is '' for both.
  // ⚠ Every value here needs `admin.requests.detail.author.<kind>` in en/ms/ta: the thread renders
  // that key from the value, so a missing one prints a raw dotted path to the requester.
  author_kind: 'ai' | 'owner' | 'org' | 'engineer'
  author_name: string
  body: string
  visibility: 'shared' | 'internal'
  awaiting_reply: boolean
  created_at: string
}

/** A screenshot attached to a request (Sprint 15.1). download_url is a signed Supabase URL or
 *  null (denied when the blob's key-org disagrees with the request's org). */
export interface OrgRequestAttachment {
  id: number
  original_filename: string
  content_type: string
  size: number
  created_at: string
  download_url: string | null
}

/**
 * The engineer's working paper behind one comment (TD-204) — **OWNER-ONLY**.
 *
 * ⚠ This never appears on an org_admin's payload. The analysis reaches the requesting organisation
 * as PROSE, through `comments`, authored `engineer`; the CITED FILES and the ENGINEER'S HOURS stay
 * owner-side. Neither is secrecy: a citation the requester cannot open buys them nothing while the
 * paths disclose the internal shape of a multi-tenant platform, and a second hours figure in front
 * of them recreates what removing the AI's estimate fixed. The field is optional on
 * `OrgRequestDetail` for exactly that reason — an org payload does not carry it at all.
 */
export interface OrgRequestAnalysis {
  id: number
  body: string
  estimated_hours: string | null
  cited_files: string[]
  authored_by: string
  repo_sha: string
  /** The engineer's PROPOSED triage. It prefills the owner's form and applies nothing — the
   *  request's own kind/lane change only when the owner presses Run. '' means "no opinion",
   *  which is not the same as agreeing with the AI draft. */
  proposed_kind: string
  proposed_lane: string
  created_at: string
  approved_at: string | null
  approved_by_name: string
  superseded_at: string | null
  /** The one the quote gate reads — server-computed, so the screen and the gate cannot disagree. */
  is_current: boolean
}

/** The ORG-facing payload (what a submitting org_admin sees) — the allowlist. NEVER carries the
 *  AI draft (ai_*) or the owner's triage. The owner (super) additionally receives the owner
 *  fields below (all optional here). */
export interface OrgRequestDetail {
  id: number
  kind: string
  title: string
  description: string
  // Optional Bugzilla-style scoping (owner-approved Sprint 15 increment); '' when unset.
  component: string
  urgency: string
  steps_to_reproduce: string
  status: string
  // The discussion, oldest first. Filtered to `shared` for an org_admin, server-side.
  comments: OrgRequestComment[]
  attachments: OrgRequestAttachment[]
  quote_hours: string | null
  quote_note: string
  quoted_at: string | null
  approved_at: string | null
  scheduled_for: string | null
  decline_reason: string
  created_at: string
  updated_at: string
  submitted_by_name: string
  // Owner-only (super payload) — undefined in the org payload by construction.
  organisation_id?: number
  organisation_name?: string
  // The margin is owner-only (owner, 2026-07-30: "do not mention the margin"). It is off the ORG
  // serializer entirely, not merely hidden in the UI, so an org payload cannot carry it.
  quote_margin_pct?: number | null
  // The engineer's analyses, newest first (TD-204). Owner-only — see OrgRequestAnalysis.
  analyses?: OrgRequestAnalysis[]
  ai_run_count?: number
  ai_draft_kind?: string
  ai_draft_lane?: string
  ai_draft_hours?: string | null
  ai_draft_note?: string
  ai_draft_model?: string
  ai_draft_at?: string | null
  triaged_kind?: string
  lane?: string
  triage_note?: string
  triaged_at?: string | null
  declined_by_role?: string
}

export async function getOrgRequests(
  filters?: { status?: string }, options?: ApiOptions
): Promise<{ requests: OrgRequestDetail[] }> {
  const qs = filters?.status ? `?status=${encodeURIComponent(filters.status)}` : ''
  return adminFetch(`/api/v1/admin/scholarship/requests/${qs}`, options)
}

export async function getOrgRequest(id: number, options?: ApiOptions): Promise<OrgRequestDetail> {
  return adminFetch(`/api/v1/admin/scholarship/requests/${id}/`, options)
}

/** The lean badge probe. 404s while REQUESTS_ENABLED is off → callers hide the hub card. */
export async function getOrgRequestCount(options?: ApiOptions): Promise<{ count: number }> {
  return adminFetch('/api/v1/admin/scholarship/requests/count/', options)
}

export async function createOrgRequest(
  data: {
    kind: string; title: string; description: string; organisation_id?: number
    component?: string; urgency?: string; steps_to_reproduce?: string
  },
  options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate('/api/v1/admin/scholarship/requests/', 'POST', data, options)
}

// Requestee (org_admin) actions
/**
 * Reply to an open question. `comment_id` names WHICH question the answer is against; omitted —
 * which is what the single reply box sends — it is the oldest open one.
 *
 * ⚠ It does NOT ration what closes. The requester speaking settles every question standing before
 * their reply, by design, so naming one never leaves the others hanging. Do not build a per-question
 * chooser on the promise that it would; it would offer a distinction the server does not make.
 *
 * ⚠ The dead `index` parameter lived here until 2026-08-18. The server renamed it `comment_id` on
 * 2026-07-31 and nothing on either side was updated, so every answer 500-ed for eighteen days.
 */
export async function answerOrgRequest(
  id: number, data: { answer: string; comment_id?: number }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/answer/`, 'POST', data, options)
}
/**
 * The OWNER asks the requester a question — the other half of the clarification thread, which
 * until 2026-07-30 only the AI could write to. Super-only server-side.
 */
export async function askOrgRequest(
  id: number, data: { question: string }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/ask/`, 'POST', data, options)
}
/**
 * Post to the DISCUSSION (TD-201) — a STATEMENT, not a question, so it awaits no reply and does
 * not spend the reviewer's question budget.
 *
 * Open to the super AND to any org_admin of the owning organisation: they can already read the
 * request, so this lets the people already in the room speak rather than only watch.
 *
 * `visibility: 'internal'` is SUPER-ONLY — the server answers 403 for anyone else, and an org
 * author can never be internal at all. Omit it for the ordinary shared case.
 */
export async function commentOrgRequest(
  id: number, data: { body: string; visibility?: 'shared' | 'internal' }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/comments/`, 'POST', data, options)
}
/**
 * Stage the engineer's analysis as a DRAFT (TD-204). Super-only; posts nothing.
 *
 * Normally written by the `record_request_analysis` management command, which additionally checks
 * that every cited path exists in the repo. This client exists so the same thing can be done from
 * the cockpit — but a citation typed here is NOT existence-checked, which is the command's whole
 * advantage.
 */
export async function recordOrgRequestAnalysis(
  id: number,
  data: { body: string; estimated_hours?: string; cited_files: string[]; authored_by?: string },
  options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/analysis/`, 'POST', data, options)
}
/** The owner approves an analysis and it enters the thread as an `engineer` comment. Super-only. */
export async function approveOrgRequestAnalysis(
  id: number, analysisId: number, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/analysis/${analysisId}/approve/`,
                     'POST', {}, options)
}
/** Retire a DRAFT analysis the engineer got wrong, so it cannot be approved by mistake. */
export async function withdrawOrgRequestAnalysis(
  id: number, analysisId: number, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/analysis/${analysisId}/withdraw/`,
                     'POST', {}, options)
}
export async function approveOrgRequest(id: number, options?: ApiOptions): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/approve/`, 'POST', {}, options)
}
export async function deferOrgRequest(id: number, options?: ApiOptions): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/defer/`, 'POST', {}, options)
}
export async function modifyOrgRequest(
  id: number, data: { description: string }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/modify/`, 'POST', data, options)
}
export async function declineOrgRequest(
  id: number, data: { reason?: string }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/decline/`, 'POST', data, options)
}

// Owner (super) actions
export async function triageOrgRequest(
  id: number, data: { triaged_kind: string; lane: string; note?: string }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/triage/`, 'POST', data, options)
}
export async function quoteOrgRequest(
  id: number, data: { hours: number; margin_pct?: number; note?: string }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/quote/`, 'POST', data, options)
}
export async function requoteOrgRequest(
  id: number, data: { hours: number; margin_pct?: number; note?: string }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/requote/`, 'POST', data, options)
}
export async function scheduleOrgRequest(
  id: number, data: { scheduled_for?: string }, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/schedule/`, 'POST', data, options)
}
export async function doneOrgRequest(id: number, options?: ApiOptions): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/done/`, 'POST', {}, options)
}
export async function aiRerunOrgRequest(id: number, options?: ApiOptions): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/ai-rerun/`, 'POST', {}, options)
}

// ---- Screenshot attachments (Sprint 15.1) ----
// The upload is the sign -> PUT bytes to Supabase -> record chain (ActionCentre.tsx:187 pattern);
// see uploadOrgRequestAttachment for the whole flow.
async function signOrgRequestAttachment(
  id: number, options?: ApiOptions
): Promise<{ upload_url: string; storage_path: string }> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/attachments/sign-upload/`, 'POST', {}, options)
}

async function recordOrgRequestAttachment(
  id: number,
  payload: { storage_path: string; original_filename: string; content_type: string; size: number },
  options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/attachments/`, 'POST', payload, options)
}

export async function deleteOrgRequestAttachment(
  id: number, attId: number, options?: ApiOptions
): Promise<OrgRequestDetail> {
  return adminMutate(`/api/v1/admin/scholarship/requests/${id}/attachments/${attId}/`, 'DELETE', null, options)
}

/** The whole upload chain: sign → PUT the bytes straight to Supabase → record the row. Returns the
 *  refreshed request (with the new attachment). Throws on any step (caller shows a warning). */
export async function uploadOrgRequestAttachment(
  id: number, file: File, options?: ApiOptions
): Promise<OrgRequestDetail> {
  const { upload_url, storage_path } = await signOrgRequestAttachment(id, options)
  const resp = await fetch(upload_url, {
    method: 'PUT',
    body: file,
    headers: { 'Content-Type': file.type || 'application/octet-stream', 'x-upsert': 'true' },
  })
  if (!resp.ok) throw new Error(`Upload failed: ${resp.status}`)
  return recordOrgRequestAttachment(id, {
    storage_path,
    original_filename: file.name,
    content_type: file.type || 'application/octet-stream',
    size: file.size,
  }, options)
}

File: halatuju_api/apps/scholarship/views_admin/requests.py
"""The Requests space — a request and its conversation. Moved verbatim from `views_admin.py`
at code health H11. The analysis, the quote and the delivery are in `requests_delivery.py`.

Part of the `views_admin` package. Every name below is re-exported from
`views_admin/__init__.py`, so `urls.py` and every importer are unchanged.
"""
import logging

from django.conf import settings
from django.db.models import Count, Q
from rest_framework import status
from rest_framework.response import Response

from ..models import OrgRequest
from ..serializers_admin import OrgRequestOrgSerializer, OrgRequestOwnerSerializer
from .base import _AdminBase

#: THE PACKAGE'S LOGGER NAME, SPELLED OUT. `__name__` here would read
#: `apps.scholarship.views_admin.requests`, and twelve tests assert on
#: `apps.scholarship.views_admin` — an audit line that moves logger is an audit line nobody
#: is watching any more. Every submodule in this package does the same.
logger = logging.getLogger('apps.scholarship.views_admin')


# ── Requests space (Sprint 15) ─────────────────────────────────────────────────────
# The org-section "Requests" area: bug/feature forms → AI reviewer → owner-gated hours
# quotes. Ships DARK behind REQUESTS_ENABLED — every route 404s while the flag is off
# (the FE hub card is hidden by the same 404-probe, so there is no client flag). Service =
# apps.scholarship.org_requests; org-fenced via _org_request_for (cross-org 404), role-gated
# per the endpoint table (org-side vs super-only). All classes classified in
# test_org_fence.py FENCED_OR_EXEMPT and the OrgRequest model is WATCHED (its raw admin
# queries below all carry an # org-fence pragma).

def _org_request_err(e):
    """Map an OrgRequestError code to a 4xx. bad_transition/bug_is_free/... are 4xx; the two
    AI-availability codes are 503 (the model is unconfigured/unavailable, not the caller's fault)."""
    if e.code in ('triage_ai_unconfigured', 'triage_ai_unavailable'):
        return Response({'error': e.code, 'code': e.code},
                        status=status.HTTP_503_SERVICE_UNAVAILABLE)
    return Response({'error': e.code, 'code': e.code}, status=status.HTTP_400_BAD_REQUEST)


class _OrgRequestsBase(_AdminBase):
    """Shared flag/role/org gate for the Requests-space endpoints.

    404-FIRST dark ship: with ``REQUESTS_ENABLED`` off, ``_flag`` short-circuits every handler to
    404 BEFORE any auth/role work — the same shape as the sponsor-pool flag gate — so the feature
    leaks no existence signal while dark. When the flag is on, role denials are REAL 403s and a
    cross-org id is 404 (no existence leak)."""

    def _flag(self):
        """Returns an error Response (404) when the feature is dark, else None."""
        if not getattr(settings, 'REQUESTS_ENABLED', False):
            return Response({'error': 'not_found'}, status=status.HTTP_404_NOT_FOUND)
        return None

    def _not_found(self):
        return Response({'error': 'not_found'}, status=status.HTTP_404_NOT_FOUND)

    def _org_request_for(self, admin, pk):
        # org-fence: fetch then re-gate to the caller's organisation (super global); a cross-org
        # id returns None -> 404 (no existence leak). This is the ONLY OrgRequest.objects read.
        req = (OrgRequest.objects
               .select_related('organisation', 'submitted_by').filter(pk=pk).first())
        if req is None:
            return None
        if self.has_role(admin, 'super'):
            return req
        if req.organisation_id != admin.owning_organisation_id:
            return None
        return req

    # ── role prologues (flag already assumed checked by the caller) ──────────────
    def _org_side(self, request):
        """Caller must be an org_admin or super (the roles that OPEN the Requests area)."""
        admin = self.get_admin(request)
        if not admin:
            return None, self._deny()
        if not (admin.is_super or admin.role == 'org_admin'):
            return None, self._deny_role()
        return admin, None

    def _requestee(self, request, pk, *, allow_super=False):
        """A requestee WRITE (answer/defer/modify → org_admin only; approve/decline → +super).
        Returns (admin, req, None) or (None, None, err)."""
        admin = self.get_admin(request)
        if not admin:
            return None, None, self._deny()
        if not ((admin.role == 'org_admin') or (allow_super and admin.is_super)):
            return None, None, self._deny_role()
        req = self._org_request_for(admin, pk)
        if req is None:
            return None, None, self._not_found()
        return admin, req, None

    def _super_side(self, request, pk):
        """A super-only WRITE (triage/quote/requote/schedule/done/ai-rerun)."""
        admin = self.get_admin(request)
        if not admin:
            return None, None, self._deny()
        if not admin.is_super:
            return None, None, self._deny_role()
        req = self._org_request_for(admin, pk)
        if req is None:
            return None, None, self._not_found()
        return admin, req, None

    def _serialize(self, admin, req):
        """Super sees the OWNER payload (incl. the AI draft + triage); everyone else the
        allowlist ORG payload (no ai_* / triage ever)."""
        if self.has_role(admin, 'super'):
            return OrgRequestOwnerSerializer(req).data
        return OrgRequestOrgSerializer(req).data


class AdminOrgRequestListView(_OrgRequestsBase):
    """GET list (org-fenced) . POST create a request. org_admin + super."""

    def get(self, request):
        gate = self._flag()
        if gate:
            return gate
        admin, err = self._org_side(request)
        if err:
            return err
        # org-fence: list scoped to the caller's organisation (super global) via _org_scoped.
        qs = self._org_scoped(
            OrgRequest.objects.select_related('organisation', 'submitted_by'),
            admin, field='organisation_id')
        return Response({'requests': [self._serialize(admin, r) for r in qs]})

    def post(self, request):
        gate = self._flag()
        if gate:
            return gate
        admin, err = self._org_side(request)
        if err:
            return err
        from .. import org_requests
        # Whose org the request belongs to: the org_admin's own; a super must name organisation_id.
        if admin.is_super:
            org_id = request.data.get('organisation_id')
            from apps.courses.models import PartnerOrganisation
            org = PartnerOrganisation.objects.filter(pk=org_id).first() if org_id else None
            if org is None:
                return Response({'error': 'org_required', 'code': 'org_required'},
                                status=status.HTTP_400_BAD_REQUEST)
        else:
            org = admin.owning_organisation
            if org is None:
                return Response({'error': 'no_org', 'code': 'no_org'},
                                status=status.HTTP_400_BAD_REQUEST)
        try:
            req = org_requests.create_request(
                org, admin, kind=(request.data.get('kind') or '').strip(),
                title=request.data.get('title') or '',
                description=request.data.get('description') or '',
                component=request.data.get('component') or '',
                urgency=request.data.get('urgency') or '',
                steps_to_reproduce=request.data.get('steps_to_reproduce') or '')
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        # Best-effort post-commit: notify the owner + auto-run the AI reviewer (never fails create).
        try:
            from .. import emails
            emails.send_org_request_submitted_email(req)
        except Exception:
            logger.warning('Requests: submit-notify failed for OrgRequest %s', req.pk, exc_info=True)
        org_requests.auto_run_ai_review(req)
        req.refresh_from_db()
        return Response(self._serialize(admin, req), status=status.HTTP_201_CREATED)


class AdminOrgRequestCountView(_OrgRequestsBase):
    """GET {count} for the nav badge. Super: requests waiting on US — SUBMITTED (awaiting triage)
    OR a triaged FEATURE with no approved analysis (TD-205). org_admin: own org's requests that
    need THEIR attention — quoted (awaiting accept) OR carrying an unanswered clarifying question.
    org_admin + super."""

    def get(self, request):
        gate = self._flag()
        if gate:
            return gate
        admin, err = self._org_side(request)
        if err:
            return err
        from django.db.models import Count, Q
        if self.has_role(admin, 'super'):
            # TD-205: "waiting on us" is BOTH ends of the engineer's involvement. Untriaged is the
            # obvious half. The other is a triaged FEATURE with no approved analysis — it cannot be
            # quoted at all (`analysis_required` refuses), so it is stuck BY CONSTRUCTION and
            # nothing else says so. A triaged BUG is deliberately NOT counted: a bug is free and
            # schedulable straight from triage, so it waits on a decision, not on an analysis.
            #
            # A filtered Count, and one annotate only (two multi-valued annotates multiply each
            # other — this project has been bitten by that).
            #
            # A single `.exclude(analyses__approved_at__isnull=False, analyses__superseded_at__
            # isnull=True)` is EQUIVALENT here and was measured to be, not assumed: Django compiles
            # one multi-condition exclude into a single NOT EXISTS with both conditions on the same
            # joined row, which is exactly "has no approved, live analysis". The multi-valued
            # negation trap is real but belongs to CHAINED `.exclude(a).exclude(b)`, which asks two
            # independent questions of two different rows. Count is kept for being explicit about
            # the zero and for not needing `.distinct()`, NOT because exclude is broken — an
            # earlier version of this comment claimed it was, and a bite-check disproved it.
            #
            # An approved analysis always carries ≥1 cited file because `approve_analysis` refuses
            # otherwise, so this agrees with `org_requests.approved_analysis` without re-testing it.
            # org-fence: super is global by design for the triage badge.
            waiting = OrgRequest.objects.annotate(
                live_analyses=Count('analyses', filter=Q(analyses__approved_at__isnull=False,
                                                         analyses__superseded_at__isnull=True)),
            ).filter(
                Q(status='submitted')
                | Q(status='triaged', triaged_kind='feature', live_analyses=0)
            )
            return Response({'count': waiting.count()})
        # TD-201: "needs you" is a quote awaiting a decision, or a question awaiting a reply —
        # the latter is now a comment row, so it is one subquery instead of walking a JSON list
        # per request.
        # org-fence: own org only (org_admin). Kept ADJACENT to the query — the static guard reads
        # a 200-char window, so an explanation wedged in between silently un-fences it.
        qs = OrgRequest.objects.filter(
            organisation_id=admin.owning_organisation_id,
        ).exclude(status__in=('done', 'declined'))
        return Response({'count': qs.filter(
            Q(status='quoted') | Q(comments__awaiting_reply=True)
        ).distinct().count()})


class AdminOrgRequestDetailView(_OrgRequestsBase):
    """GET one request (org_admin own else 404; super). org_admin + super."""

    def get(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, err = self._org_side(request)
        if err:
            return err
        req = self._org_request_for(admin, pk)
        if req is None:
            return self._not_found()
        return Response(self._serialize(admin, req))


class AdminOrgRequestAnswerView(_OrgRequestsBase):
    """POST answer a clarifying question (org_admin own org). No status transition."""

    def post(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, req, err = self._requestee(request, pk)
        if err:
            return err
        from .. import org_requests
        try:
            # ⚠ `comment_id` AND `admin`, and both were missing. This call still passed `index=`
            # — the parameter the service dropped on 2026-07-31 when clarifications became
            # comments — so EVERY answer raised TypeError before the service was reached, and the
            # `except OrgRequestError` below could not see it. Answering was 500-ing for every
            # organisation on every request for eighteen days (BrightPath request #15).
            req = org_requests.answer_clarification(
                req, request.data.get('answer') or '',
                comment_id=request.data.get('comment_id'),
                admin=admin)
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        # Best-effort: notify the owner + re-run the AI reviewer on the new answer.
        try:
            from .. import emails
            emails.send_org_request_answered_email(req)
        except Exception:
            logger.warning('Requests: answer-notify failed for OrgRequest %s', req.pk, exc_info=True)
        org_requests.auto_run_ai_review(req)
        req.refresh_from_db()
        return Response(self._serialize(admin, req))


class AdminOrgRequestApproveView(_OrgRequestsBase):
    """POST accept a quote (quoted/deferred → approved). org_admin own org, or super."""

    def post(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, req, err = self._requestee(request, pk, allow_super=True)
        if err:
            return err
        from .. import org_requests
        by_role = 'super' if admin.is_super else 'org_admin'
        try:
            req = org_requests.approve(req, admin, by_role=by_role)
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        try:
            from .. import emails
            emails.send_org_request_accepted_email(req)
        except Exception:
            logger.warning('Requests: accept-notify failed for OrgRequest %s', req.pk, exc_info=True)
        return Response(self._serialize(admin, req))


class AdminOrgRequestDeferView(_OrgRequestsBase):
    """POST defer a quote (quoted → deferred). org_admin own org."""

    def post(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, req, err = self._requestee(request, pk)
        if err:
            return err
        from .. import org_requests
        try:
            req = org_requests.defer(req, admin)
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        return Response(self._serialize(admin, req))


class AdminOrgRequestModifyView(_OrgRequestsBase):
    """POST modify (amend the description; quoted/deferred → submitted). org_admin own org."""

    def post(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, req, err = self._requestee(request, pk)
        if err:
            return err
        from .. import org_requests
        try:
            req = org_requests.modify(req, admin, description=request.data.get('description') or '')
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        org_requests.auto_run_ai_review(req)
        req.refresh_from_db()
        return Response(self._serialize(admin, req))


class AdminOrgRequestDeclineView(_OrgRequestsBase):
    """POST decline/withdraw (→ declined, terminal). org_admin own org (withdraw, reason
    optional), or super (decline, reason required)."""

    def post(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, req, err = self._requestee(request, pk, allow_super=True)
        if err:
            return err
        from .. import org_requests
        by_role = 'super' if admin.is_super else 'org_admin'
        try:
            req = org_requests.decline(req, admin, by_role=by_role,
                                       reason=request.data.get('reason') or '')
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        return Response(self._serialize(admin, req))


class AdminOrgRequestAskView(_OrgRequestsBase):
    """POST <pk>/ask/ {question} — the OWNER asks the requester something. Super only.

    Until now the clarification thread ran one way: the AI asked, the requester answered, and the
    owner watched by email. So a judgement about the SHAPE of a request — "adding a sponsor
    directly would bypass the terms and consent; would an invite do?" — had nowhere to go, because
    `triage_note` is private to the owner and the org never sees it.

    Same window as `/answer/` and the AI's own questions (submitted/triaged): a quoted request
    must not grow new questions, because the quote was priced against what was known when it
    was sent.

    Emails the requester through the SAME helper the AI's questions use, so a question reads the
    same to them however it was authored — only the on-screen attribution differs.
    """

    def post(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, req, err = self._super_side(request, pk)
        if err:
            return err
        from .. import org_requests
        try:
            question = org_requests.ask_question(req, admin, request.data.get('question') or '')
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        try:
            from .. import emails
            emails.send_org_request_questions_email(req, [question])
        except Exception:
            logger.warning('Requests: owner-question notify failed for OrgRequest %s',
                           req.pk, exc_info=True)
        req.refresh_from_db()
        return Response(self._serialize(admin, req))


class AdminOrgRequestCommentView(_OrgRequestsBase):
    """POST <pk>/comments/ {body, visibility?} — post to the DISCUSSION (TD-201).

    The verb the module never had. Until now exactly ONE action reached the requester: `ask` a
    question. So a conclusion — "here is what we would build, and why" — had to travel as a quote
    note or not at all, and the owner's judgement about the shape of a request left the system.

    ACTOR: super OR any org_admin of the owning organisation (owner ruling, 2026-07-31). They can
    already READ the request — requests are org-fenced, and a cross-org pk is a 404 — so this adds
    no visibility, it lets the people already in the room speak. `_requestee(allow_super=True)` is
    exactly that rule; the org fence is the request lookup, not a check here.

    ⚠ `visibility='internal'` is SUPER-ONLY and the service refuses it for an org author. Two
    layers on purpose: a serializer allowlist cannot save you here, because the leak would be a
    ROW the org may not read rather than a field — see `org_requests.comments_for`.

    WINDOW: until the request is TERMINAL, wider than `OPEN_FOR_SHAPING`. Discussion continues
    after assignment (the owner's Bugzilla framing); it is asking a NEW QUESTION that still stops
    at the quote, because a question can re-price and a remark cannot.
    """

    def post(self, request, pk):
        gate = self._flag()
        if gate:
            return gate
        admin, req, err = self._requestee(request, pk, allow_super=True)
        if err:
            return err
        from .. import org_requests
        visibility = (request.data.get('visibility') or org_requests.VISIBILITY_SHARED).strip()
        # An org_admin may not post an internal note. Refused HERE as well as in the service so
        # the endpoint's contract is readable without following the call.
        if visibility == org_requests.VISIBILITY_INTERNAL and not admin.is_super:
            return Response({'error': 'forbidden', 'code': 'forbidden'},
                            status=status.HTTP_403_FORBIDDEN)
        author_kind = (org_requests.AUTHOR_OWNER if admin.is_super
                       else org_requests.AUTHOR_ORG)

        # ⚠ THE ENGINEER MAY SPEAK DIRECTLY, BUT ONLY WHERE THE ORGANISATION CANNOT HEAR IT
        # (2026-08-01). Authorship is otherwise derived from the caller, which meant a note the
        # ENGINEER wrote — a triage recommendation, say — arrived stamped as the OWNER, because it
        # is the owner's token making the call. TD-204 already refused that trade for approved
        # analyses ("attributing it to the approver is a lie about who wrote it"); the same
        # objection applies to a note the owner did not write.
        #
        # ⚠ INTERNAL ONLY, and the pairing is the whole control. Engineer prose that REACHES the
        # requester still has exactly one route — stage an analysis, the owner approves — so this
        # cannot become a side door around that gate. An internal note is owner-visible by
        # construction (`org_requests.comments_for` filters the ROW), so there is nothing for an
        # approval step to protect.
        #
        # ⚠ THE RULE LIVES HERE, NOT IN `post_comment`, and that is deliberate rather than lazy:
        # `approve_analysis` legitimately posts engineer + SHARED through the same service, so a
        # service-level "engineer implies internal" would break the one path this exists to
        # protect. What is enforced here is the HTTP contract — who may claim to be whom — while
        # the domain rule (engineer + shared happens only on approval) stays in the service.
        claimed = (request.data.get('author') or '').strip()
        if claimed:
            if not admin.is_super or claimed != org_requests.AUTHOR_ENGINEER:
                return Response({'error': 'forbidden', 'code': 'forbidden'},
                                status=status.HTTP_403_FORBIDDEN)
            if visibility != org_requests.VISIBILITY_INTERNAL:
                return Response({'error': 'engineer_must_be_internal',
                                 'code': 'engineer_must_be_internal'},
                                status=status.HTTP_400_BAD_REQUEST)
            author_kind = org_requests.AUTHOR_ENGINEER
        try:
            org_requests.post_comment(
                # ⚠ `author_admin=None` for the engineer, exactly as `approve_analysis` does.
                # `_comment_dicts` exposes `author_name`, so passing the calling admin would print
                # the OWNER'S NAME beside an "Engineer" badge — the same lie in a second field,
                # and the one a reader would actually see.
                req, None if author_kind == org_requests.AUTHOR_ENGINEER else admin,
                request.data.get('body') or '',
                author_kind=author_kind, visibility=visibility)
        except org_requests.OrgRequestError as e:
            return _org_request_err(e)
        req.refresh_from_db()
        return Response(self._serialize(admin, req))
